### PR TITLE
Fix the carryall freeze and untangle the afterLandActivity mess

### LINKED ISSUE #14572: Carryalls in Dune 2000 crash the game by getting stuck in a loop
It somehow loops while setting the parent activity. See [this video](https://www.dropbox.com/s/un4g9prj3s0sumy/dune%20carryall%20crash.flv?dl=0) on how to easily reproduce the crash.

### PR DESCRIPTION
**Pull request by abcdefg30**
Fixes #14572.
Supersedes #14595.

This PR removes `MovingToResources`, `MovingToRefinery` and `MovementCancelled` from the `INotifyHarvesterAction` interface, as those methods are redundant to what `ICallForTransport` offers.
We also avoid getting stuck in a loop by not creating a new activity and queuing another new one as `landAfterActivity` right afterwards.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Should be fixed now.

**Comment by GraionDilach** [comment]
This literally breaks the RA2 Chrono harvester.

**Comment by abcdefg30** [comment]
Hm, one could work around that by using the `ICallForTransport` interface, that's not very nice though. On a quick thought: What about having `Harvester` + its activities only care about `INotifyHarvesterAction` (without removing the extra interface methods) and having `AutoCarryable` only care about `ICallForTransport` (fulfilling the second request behind this PR) with a third trait bridging harvester actions and transport requests?

**Comment by reaperrr** [comment]
> with a third trait bridging harvester actions and transport requests?

What about a `CallsForTransport` bool flag on `Harvester` instead?

Your proposal might be more sensible, right now I just can't picture how that bridging trait would look like (let alone how to call it).

**Comment by Unit158** [comment]
This should be labeled "Unaddressed comments" not "Unadressed comments"

**Comment by pchote** [comment]
@GraionDilach's point about the chrono harvester invalidates the original motivation for removing `INotifyHarvesterAction.MovingToRefinery` etc.  Considering that, keeping `AutoCarryable : INotifyHarvesterAction` is IMO the most sensible implementation.  Can we revert that change from here?

**Comment by GraionDilach** [comment]
In that case, I would actually also file an issue about axing the interfaces, instantly assigned to the Future milestone with the flashed out locomotor setup being it's prerequisite - technically, the RA2 harvester could jump to any dock (service depot included) and it would make sense to shift the implementation to a locomotor longterm instead of what it is atm.

**Comment by pchote** [comment]
Ping @abcdefg30.  It would be good to get some progress here both to fix this crash finally for the next release, and also to help clean up the PR queue :smile:

**Comment by pchote** [comment]
If we can get this ready and reviewed in time for the playtest, then great.  I don't think we can justify delaying the playtest much more specifically for this, however.  Bumping to Next + 1.

**Comment by abcdefg30** [comment]
Rebased & updated.

Added a new `CarryableHarvester` trait that does the bridging between `INotifyHarvesterAction` and `ICallForTransport`. (That was previously done in `AutoCarryable`, so that trait is "untangled" now as well.)

**Comment by GraionDilach** [comment]
This will need an upgrade rule?

**Comment by abcdefg30** [comment]
Yes, that might be a good idea.

**Comment by abcdefg30** [comment]
Rebased and added a rule.

**Comment by pchote** [comment]
Needs a rebase, and the update rule should move to the new directory.

**Comment by abcdefg30** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 1]
```diff
@@ -233,11 +233,8 @@ public void UnblockRefinery(Actor self)
 					var unblockCell = LastHarvestedCell ?? (deliveryLoc + Info.UnblockCell);
 					var moveTo = mobile.NearestMoveableCell(unblockCell, 1, 5);
 
-					// TODO: The harvest-deliver-return sequence is a horrible mess of duplicated code and edge-cases
-					var notify = self.TraitsImplementing<INotifyHarvesterAction>();
-					var findResources = new FindResources(self);
-					foreach (var n in notify)
-						n.MovingToResources(self, moveTo, findResources);
+					// FindResources cares about calling INotifyHarvesterAction
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: 'cares about' -> 'takes care of'

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 2]
```diff
@@ -381,12 +378,10 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 					loc = self.Location;
 				}
 
-				var findResources = new FindResources(self);
-				self.QueueActivity(findResources);
 				self.SetTargetLine(Target.FromCell(self.World, loc.Value), Color.Red);
 
-				foreach (var n in notify)
-					n.MovingToResources(self, loc.Value, findResources);
+				// FindResources cares about calling INotifyHarvesterAction
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: 'cares about' -> 'takes care of'

**Review by reaperrr — DISMISSED** [review]

**Comment by reaperrr** [comment]
Otherwise looks good to me, needs a rebase as well, though.

**Comment by abcdefg30** [comment]
Updated.

**Comment by pchote** [comment]
I have pushed an update to this branch:
 * Fix a typo in the second commit's message
 * Rebase
 * Fix the update rule to not refer to a specific line number

**Review by pchote — APPROVED** [review]
Spectated a very long D2K AI game without any noticable issues. Code LGTM.

**Pull request merged**